dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,257 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import traceback
+import os
+import socket
+from multiprocessing import reduction
+from nvidia.dali.shared_batch import SharedMemChunk, write_batch
+from nvidia.dali.messages import CompletedTasks
+
+
+class _ProcessedTasks:
+    """Internal worker message send to disptacher with completed tasks where it is
+    serialized and dispatched to the pool"""
+
+    def __init__(self, scheduled, mem_chunk=None, data_batch=None, exception=None,
+                 traceback_str=None):
+        self.context_i = scheduled.context_i
+        self.batch_i = scheduled.batch_i
+        self.mem_chunk = mem_chunk
+        self.data_batch = data_batch
+        self.exception = exception
+        self.traceback_str = traceback_str
+
+    @classmethod
+    def done(cls, scheduled, mem_chunk, data_batch):
+        return cls(scheduled, mem_chunk, data_batch)
+
+    @classmethod
+    def failed(cls, scheduled, exception, traceback_str=None):
+        return cls(scheduled, exception=exception, traceback_str=traceback_str)
+
+    def is_failed(self):
+        return self.exception is not None
+
+
+class SharedBatchesDispatcher:
+    """SharedBatchesDispatcher serializes batches, puts them into provided
+    shared memory chunks and notifies parent process of batch ready to be read from shared memory.
+    It keeps track of what shared memory chunks have been already sent and if needed, sends
+    file descriptors of memory chunks that parent process hasn't seen yet.
+
+    Parameters
+    ----------
+    `worker_id` : int
+        Id of the worker passed by the parent process. Added to messages sent over the ``res_pipe``
+        to simplify bookeeping in parent process.
+    `sock` : socket
+        Python wrapper around Unix socket, capable of sending file descriptors between processes.
+    `res_pipe`: pipe
+        Pipe used to send parent process a notification (along with essential meta data info) about
+        ready batch in a given shared memory chunk.
+    """
+
+    def __init__(self, worker_id, sock, res_pipe):
+        self.worker_id = worker_id
+        self.handle_sent = set()
+        self.sock = sock
+        self.res_pipe = res_pipe
+        self.ready_cv = threading.Condition()
+        self.ready_queue = []
+
+    def dispatch(self, processed_task: _ProcessedTasks):
+        """Pass the processed task (or None to end) to the dispatcher.
+        """
+        with self.ready_cv:
+            if processed_task is None:
+                self.ready_queue.insert(0, None)
+            else:
+                # assert len(ready_queue) < prefetch_queue_depths[scheduled.context_i], "Worker queue size exceeded."
```

Comment:
?

Reply:
Hmm, yeah, the assert was there before refactor, and after refactor I got encapsulated out of the data.

Reply:
TBH, it doesn't make much sense, as the queue of the thread that sends back the data can be longer than the prefetch_queue_depth for given callback (as it can have accumulated data from several callbacks).

Reply:
So maybe just remove it?

Reply:
Removed.